### PR TITLE
Alerting: Add file provisioning for alert rules

### PR DESCRIPTION
**Pull request by JohnnyQQQQ**
This PR introduces file provisioning for alert rules.

Note for the reviewer: 

- We support both, yaml and json. The yaml parser can also handle json, that's why you will only see a call for the yaml parser.
- I added in the provisioning dir the alerting folder that will not only hold the rules folder but also all the other resources that we will add later as contact points etc.
- Long term we should add our alerting provisioning services to wire.go but I think it is out of scope for this pr.

### CODE REVIEW

File: conf/provisioning/alertrules/rules2.yml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      labels:
+        team: infra
+        severity: warning
+      data:
+      - refId: A
+        queryType: ''
+        relativeTimeRange:
+          from: 600
+          to: 0
+        datasourceUid: PD8C576611E62080A
```
Comment (alexweav) [inline conf/provisioning/alertrules/rules2.yml:22]: Odd question - when you provision a dashboard, how does it handle datasourceUID? Presumably datasourceUIDs will be different on different instances? How does this not break, does it have a system for pointing to another resource?

Comment (alexweav) [reply conf/provisioning/alertrules/rules2.yml:22]: If dashboards has a mechanism for this, let's lift it

Comment (JohnnyQQQQ) [reply conf/provisioning/alertrules/rules2.yml]: Talked about this offline. UID should be user provided.

File: pkg/services/provisioning/alerting/rules/provisioner.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	return &DefaultAlertRuleProvisioner{
+		logger:               logger,
+		cfgReader:            newRulesConfigReader(logger),
+		dashboardService:     dashboardService,
+		dashboardProvService: dashboardProvService,
+		ruleService:          ruleService,
+	}
+}
+
+type DefaultAlertRuleProvisioner struct {
```
Comment (alexweav) [inline pkg/services/provisioning/alerting/rules/provisioner.go:34]: Can this be non-exported?

File: pkg/services/provisioning/alerting/rules/provisioner.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+func Provision(
+	ctx context.Context,
+	path string,
+	dashboardService dashboards.DashboardService,
+	dashboardProvisioningService dashboards.DashboardProvisioningService,
+	ruleService provisioning.AlertRuleService,
+) error {
+	ruleProvisioner := NewAlertRuleProvisioner(
+		log.New("provisioning"),
```
Comment (alexweav) [inline pkg/services/provisioning/alerting/rules/provisioner.go:50]: We already created a logger with the same name in `provisioning.go`. Should we rename this, or just inject that logger?

File: pkg/services/provisioning/alerting/rules/types.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+type RuleDeleteV1 struct {
+	UID   values.StringValue `json:"uid" yaml:"uid"`
+	OrgID values.Int64Value  `json:"orgId" yaml:"orgId"`
+}
+
+type AlertRuleGroupV1 struct {
+	OrgID    values.Int64Value  `json:"orgId" yaml:"orgId"`
+	Name     values.StringValue `json:"name" yaml:"name"`
+	Folder   values.StringValue `json:"folder" yaml:"folder"`
```
Comment (alexweav) [inline pkg/services/provisioning/alerting/rules/types.go:67]: This has to be a UID right? Should we call this `FolderUID` to reduce confusion? I could see users trying to put a folder name in here and getting confused why it isn't working.

File: pkg/services/provisioning/alerting/rules/provisioner.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	_, _, err := prov.ruleService.GetAlertRule(ctx, orgID, rule.UID)
+	if err != nil && !errors.Is(err, alert_models.ErrAlertRuleNotFound) {
+		return err
+	} else if err != nil {
+		prov.logger.Debug("creating rule", "uid", rule.UID, "org", rule.OrgID)
+		_, err = prov.ruleService.CreateAlertRule(ctx, rule, alert_models.ProvenanceFile)
+	} else {
+		prov.logger.Debug("updating rule", "uid", rule.UID, "org", rule.OrgID)
+		_, err = prov.ruleService.UpdateAlertRule(ctx, rule, alert_models.ProvenanceFile)
+	}
```
Comment (alexweav) [inline pkg/services/provisioning/alerting/rules/provisioner.go:127]: I'm close to PR'ing an internal API which just sets a whole group. That would greatly simplify this logic and reduce the number of database calls.

If that PR has not gone in when this one does, ping me. I will refactor this to use it.

**Review by alexweav — APPROVED** [review]
Handful of nits, nothing serious. LGTM

File: pkg/services/provisioning/alerting/rules/types.go (hunk 4 above)
Comment (alexweav) [reply pkg/services/provisioning/alerting/rules/types.go:67]: I realized this was wrong after the fact. It's folder name.

File: pkg/services/provisioning/alerting/rules/provisioner.go (hunk 3 above)
Comment (JohnnyQQQQ) [reply pkg/services/provisioning/alerting/rules/provisioner.go:50]: i changed this to `provisioning.alerting`, it's the same format as the others have.

File: pkg/services/provisioning/provisioning.go [hunk 6]
```diff
@@ -218,6 +240,28 @@ func (ps *ProvisioningServiceImpl) ProvisionDashboards(ctx context.Context) erro
 	return nil
 }
 
+func (ps *ProvisioningServiceImpl) ProvisionAlertRules(ctx context.Context) error {
+	alertRulesPath := filepath.Join(ps.Cfg.ProvisioningPath, "alertrules")
```
[the remark this answers is not in the document]
Comment (JohnnyQQQQ) [reply pkg/services/provisioning/provisioning.go:244]: Talked about this offline, it's a nice idea that we will implement.

**Pull request merged**